### PR TITLE
[New] Drop Queue Window Scheduler

### PR DESCRIPTION
**Pull request by AlexanderKanakis**
For issue: https://github.com/WideChat/Rocket.Chat/issues/1225

-Activates 120 second timer when queue window is set as active which drops the queue window.
-Resets timeout to 120 seconds if DF sends payload with `continue_blackout` parameter.

### CODE REVIEW

File: enum/Scheduler.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 export enum JobName {
     SESSION_MAINTENANCE = 'session-maintenance',
     EVENT_SCHEDULER =  'event-scheduler',
+    QUEUE_WINDOW_SCHEDULER =  'queue-window-scheduler',
```
Comment (Shailesh351) [inline enum/Scheduler.ts:4]: Can we rename it to something relatable? How about `CONTINUE_BLACKOUT_TIMEOUT_SCHEDULER`?

File: lib/QueueWindowProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    public async processor(jobContext: IJobContext, read: IRead, modify: IModify, http: IHttp, persistence: IPersistence): Promise<void> {
+        const sessionId = jobContext.rid;
+        try {
+            await setIsQueueWindowActive(read, persistence, sessionId, false);
+        } catch (error) {
+            // Failed to send event, so close blackout window
+            console.error(`${Logs.DIALOGFLOW_REST_API_ERROR}: { roomID: ${sessionId} } ${getError(error)}`);
+        }
+
+    }
```
Comment (Shailesh351) [inline lib/QueueWindowProcessor.ts:23]: You are trying to set`setIsQueueWindowActive` flag to `false` But the actual flag that blocks user messages is `setIsProcessingMessage` So we need to use that maybe?

File: lib/payloadAction.ts [hunk 3]
```diff
@@ -74,6 +74,17 @@ export const  handlePayloadActions = async (app: IApp, read: IRead,  modify: IMo
...
                         if (params.continue_blackout) {
+                            const continueBlackoutTask = {
+                                id: JobName.QUEUE_WINDOW_SCHEDULER,
+                                when: `120 seconds`,
+                                data: {
+                                    rid,
+                                    sessionId: rid,
+                                    jobName: JobName.EVENT_SCHEDULER,
+                                },
+                            };
```
Comment (Shailesh351) [inline lib/payloadAction.ts:85]: Here, the `id` is `JobName.QUEUE_WINDOW_SCHEDULER,` but the `jobName` is `JobName.EVENT_SCHEDULER`. It should be same right?

File: lib/payloadAction.ts [hunk 4]
```diff
@@ -74,6 +74,17 @@ export const  handlePayloadActions = async (app: IApp, read: IRead,  modify: IMo
 
                         // Start blackout window
                         if (params.continue_blackout) {
+                            const continueBlackoutTask = {
+                                id: JobName.QUEUE_WINDOW_SCHEDULER,
+                                when: `120 seconds`,
```
Comment (Shailesh351) [inline lib/payloadAction.ts:79]: In the past we faced some issue with passing literal strings here. It was scheduling job with some invalid date. And using Date object works fine https://github.com/WideChat/Apps.Dialogflow/blob/master/lib/payloadAction.ts#L63 Can you please try it?

File: lib/EventTimeoutProcessor.ts [hunk 5]
```diff
@@ -46,6 +57,7 @@ export class EventScheduler implements IProcessor {
             const queuedMessage = await getQueuedMessage(read, sessionId);
 
             await setIsQueueWindowActive(read, persistence, sessionId, false);
+            await modify.getScheduler().cancelJobByDataQuery({sessionId, jobName: JobName.QUEUE_WINDOW_SCHEDULER});
```
Comment (Shailesh351) [inline lib/EventTimeoutProcessor.ts:60]: We are setting and unsetting `setIsQueueWindowActive` in this function. The actual flag that blocks user messages is `setIsProcessingMessage` So maybe we don't need this here?

**Comment by ear-dev** [comment]
@Shailesh351 it looks like Alexander responded to the comments?  Please verify when you have a chance.  Thanks.

**Review by Shailesh351 — APPROVED** [review]

**Pull request merged**